compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirNotImplementedOverrideChecker.kt

```diff
@@ -0,0 +1,287 @@
+import org.jetbrains.kotlin.fir.symbols.CallableId
+import org.jetbrains.kotlin.fir.symbols.impl.*
+import org.jetbrains.kotlin.fir.types.*
+import org.jetbrains.kotlin.name.Name
+
+object FirNotImplementedOverrideChecker : FirClassChecker() {
+    // TODO: differentiate members with different annotations, e.g., @Api...(x) @Api...(y)
+    // TODO: differentiate parameter names? e.g., foo(x : T) v.s. override foo(y : T)
+    private object SignaturePresenter : FirDeclarationPresenter {
+        override fun StringBuilder.appendRepresentation(it: CallableId) {
+            append(it.callableName)
+        }
+    }
+
+    private fun FirClass<*>.unsubstitutedScope(context: CheckerContext) =
+        this.unsubstitutedScope(context.sessionHolder.session, context.sessionHolder.scopeSession, withForcedTypeCalculator = true)
+
+    // TODO: Collection extensions are not properly regarded as an override. Should not have a whitelist like this.
+    private val COLLECTIONS = listOf("java.util", "kotlin.collections")
+
+    override fun check(declaration: FirClass<*>, context: CheckerContext, reporter: DiagnosticReporter) {
+        // TODO: kt4763Property: reporting on `object` literal causes invalid error in test...FirDiagnosticHandler
+        if (declaration is FirAnonymousObject) return
+
+        val source = declaration.source ?: return
+        if (source.kind is FirFakeSourceElementKind) return
+
+        val contributedMembers = collectCallableMembers(declaration, context)
+        val potentialFakeOverrides = getBaseDeclarationsForFakeOverrides(declaration, contributedMembers, context)
+
+        // TODO: consider open as overridable, and check conflicts as well
+        val abstractFakeOverrides = potentialFakeOverrides.filter {
+            it is FirMemberDeclaration &&
+                    it.isAbstract &&
+                    // TODO: FIR MPP support
+                    (it.getContainingClass(context) as? FirRegularClass)?.isExpect == false
+        }
+        if (abstractFakeOverrides.isEmpty()) return
+
+        // TODO: differentiate type-substituted declarations. Otherwise, they will be reported as MANY_IMPL_MEMBER_NOT_IMPLEMENTED
+        val sigToDeclarations = mutableMapOf<String, MutableList<FirCallableDeclaration<*>>>()
+        for (abstractFakeOverride in abstractFakeOverrides) {
+            val sig = when (abstractFakeOverride) {
+                is FirSimpleFunction -> SignaturePresenter.represent(abstractFakeOverride)
+                is FirProperty -> SignaturePresenter.represent(abstractFakeOverride)
+                else -> continue
+            }
+            sigToDeclarations.computeIfAbsent(sig) { mutableListOf() }.add(abstractFakeOverride)
+        }
+
+        val canHaveAbstractMembers = declaration is FirRegularClass && declaration.canHaveAbstractFakeOverride
+        // var alreadyReportedManyNotImplemented = false
+        var alreadyReportedAbstractNotImplemented = false
```

Comment:
Can you only report `*_NOT_IMPLEMENTED` once per class? What if there are multiple members not implemented?

Reply:
That is a general question to diagnostic framework, but for now, a single source can be used by multiple diagnostics.

Indeed, the main purpose of this flag is to _not_ report same ABSTRACT_*_NOT_IMPLEMENTED diagnostics on a class that doesn't implement multiple abstract members.

Reply:
That is interesting. I verified that just now in IntelliJ and it looks like only 1 ABSTRACT_*_NOT_IMPLEMENTED error is reported. But the quickfix to implement missing members collects all of the missing members. Something to consider for the future is to have some way to share the code between checker and the quickfix. FE1.0 seems to have this already in `OverrideResolver` class which is used in both checking and the quickfix.

Reply:
Yeah, agree. IIRC, `OverrideResolver` has a util that returns missing members, and indeed that makes sense to share the same logic between checker and quickfix. I still have checker's mind set maybe, but I believe it won't be that difficult to utilize how this checker is retrieving (pretended) fake overrides (which I believe quickfix will need too).